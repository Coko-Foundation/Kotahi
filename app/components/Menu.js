import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { th, grid } from '@pubsweet/ui-toolkit'
import { Link, useLocation } from 'react-router-dom'
import { Icon } from '@pubsweet/ui'
import { UserAvatar } from './component-avatar/src'
import lightenBy from '../shared/lightenBy'

const Root = styled.nav`
  background: linear-gradient(
    134deg,
    ${th('colorPrimary')},
    ${lightenBy('colorPrimary', 0.3)}
  );
  border-right: 1px solid ${th('colorFurniture')};
  grid-area: menu;
  max-height: 100vh;
  padding: ${grid(2)};
`

const UserName = styled.div`
  word-break: break-word;
`

const Section = styled.div``

const NavItem = ({ className, link, name, icon }) => (
  <Link className={className} to={link}>
    <Icon>{icon}</Icon>
    {name}
  </Link>
)

NavItem.propTypes = {
  className: PropTypes.string.isRequired,
  link: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  icon: PropTypes.string.isRequired,
}

export const Item = styled(NavItem)`
  align-items: center;
  background-color: ${props =>
    props.active ? th('colorBackgroundHue') : 'unset'};
  border-radius: 10px;
  color: ${props => (props.active ? th('colorText') : th('colorTextReverse'))};
  display: flex;
  height: ${grid(5)};
  line-height: ${grid(3)};
  padding-left: ${grid(1)};

  svg {
    stroke: ${props =>
      props.active ? th('colorText') : th('colorTextReverse')};
    width: 1em;
  }

  &:hover {
    background-color: ${lightenBy('colorPrimary', 0.5)};
    color: ${th('colorText')};
    stroke: ${th('colorText')};

    svg {
      stroke: ${th('colorText')};
    }
  }
`

const UserItem = styled(Link)`
  color: ${th('colorTextReverse')};
  display: flex;
  padding-bottom: ${grid(2)};
`

const UserInfo = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-left: ${grid(1)};
`

const Menu = ({
  className,
  loginLink = '/login',
  navLinkComponents,
  user,
  notice,
  profileLink,
}) => {
  const location = useLocation()
  return (
    <Root className={className}>
      <Section>
        {/* TODO: Place this notice (used for offline notification) better */}
        {notice}
        <UserComponent
          loginLink={loginLink}
          profileLink={profileLink}
          user={user}
        />
        {navLinkComponents &&
          navLinkComponents.map((navInfo, idx) => (
            <Item
              {...navInfo}
              active={location.pathname === navInfo.link}
              key={navInfo.link}
            />
          ))}
      </Section>
    </Root>
  )
}

const UserComponent = ({ user, loginLink, profileLink }) => (
  <Section>
    {user && (
      <UserItem title="Go to your profile" to={profileLink}>
        <UserAvatar isClickable={false} size={48} user={user} />
        <UserInfo>
          <UserName>{user.username}</UserName>
          <p>{user.online ? '' : 'Offline'}</p>
          {/* ({user.username}) */}
          {user.admin ? ' (admin)' : ''}
        </UserInfo>
      </UserItem>
    )}
    {!user && <Item icon="logIn" link={loginLink} name="Login" />}
  </Section>
)

Menu.propTypes = {
  className: PropTypes.string.isRequired,
  loginLink: PropTypes.string.isRequired,
  navLinkComponents: PropTypes.arrayOf(PropTypes.object).isRequired,
  user: PropTypes.oneOfType([PropTypes.object]),
  notice: PropTypes.node.isRequired,
  profileLink: PropTypes.string.isRequired,
}

Menu.defaultProps = {
  user: undefined,
}

UserComponent.propTypes = {
  user: PropTypes.oneOfType([PropTypes.object]),
  loginLink: PropTypes.string.isRequired,
  profileLink: PropTypes.string.isRequired,
}

UserComponent.defaultProps = {
  user: undefined,
}

export default Menu
