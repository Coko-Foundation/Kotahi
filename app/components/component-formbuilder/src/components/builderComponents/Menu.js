import React from 'react'
import { Select, TextField, ValidatedFieldFormik } from '@pubsweet/ui'
import { Legend, Section } from '../style'

const ValidationMenu = input => (
  // const validations = useState([])

  <>
    <Select
      {...input}
      // onChange={select => input.onChange(select.map(s => s.value))}
      // onChange={select => setValidations(select.map(s => s.value))}
    />

    {((Array.isArray(input.value) && input.value) || []).map(validation => {
      if (validation.value !== 'required') {
        return (
          <Section key={validation.value}>
            <Legend space>{validation.label} value</Legend>
            <ValidatedFieldFormik
              component={TextField}
              name={`validateValue.${validation.value}`}
            />
          </Section>
        )
      }

      return null
    })}
  </>
)

export default ValidationMenu
// export default compose(
//   withState('selectelement', 'changeSelect', undefined),
//   withHandlers({
//     onSelectElement: ({ changeSelect }) => value => changeSelect(() => value),
//   }),
// )(ValidationMenu)
