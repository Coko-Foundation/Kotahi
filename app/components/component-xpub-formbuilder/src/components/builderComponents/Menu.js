import React from 'react'
import { Menu, TextField, ValidatedFieldFormik } from '@pubsweet/ui'
import { compose, withState, withHandlers } from 'recompose'
import { Legend, Section } from '../styles'

const ValidationMenu = input => (
  <div>
    <Menu
      {...input}
      selectElement={value => {
        input.onSelectElement(value)
      }}
    />
    {input.selectelement && input.selectelement !== 'required' && (
      <Section>
        <Legend space>FIeld Min / Max</Legend>
        <ValidatedFieldFormik
          component={TextField}
          name={`validateValue.${input.selectelement}`}
        />
      </Section>
    )}
  </div>
)

export default compose(
  withState('selectelement', 'changeSelect', undefined),
  withHandlers({
    onSelectElement: ({ changeSelect }) => value => changeSelect(() => value),
  }),
)(ValidationMenu)
