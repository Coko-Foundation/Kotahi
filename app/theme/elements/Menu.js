import { css } from 'styled-components'

export default {
  // TODO
  // -- input padding: breaking the grid?
  // -- small placeholder text? maybe by default?

  Opener: css`
    height: auto;
  `,
}
