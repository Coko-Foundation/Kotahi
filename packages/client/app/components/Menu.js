/* eslint-disable react/jsx-handler-names */
import React, { useState, useMemo } from 'react'
import PropTypes from 'prop-types'
import { useTranslation } from 'react-i18next'
import { useMutation } from '@apollo/client'
import {
  NavItem,
  NavLinks,
  MainNavWrapper,
  SectionNavLayoutSettings,
  Root,
  ScrollWrapper,
  SubMenu,
  UserComponent,
  StyledPinButton,
} from './component-menu'
import mutations from './component-dashboard/src/graphql/mutations'

const Menu = ({
  className,
  loginLink = '/login',
  navLinkComponents,
  user,
  notice,
  profileLink,
}) => {
  const { t } = useTranslation()
  const [menuIsMinimal, setMenuIsMinimal] = useState(!user.menuPinned)
  const [menuPinned, setMenuPinned] = useState(!!user.menuPinned)
  const [updateMenuState] = useMutation(mutations.updateMenu)

  const renderLinks = useMemo(() => {
    return (
      navLinkComponents &&
      navLinkComponents.map(navInfo =>
        navInfo.menu ? (
          <SubMenu
            key={navInfo.name}
            menuIsMinimal={menuIsMinimal}
            setMenuIsMinimal={setMenuIsMinimal}
            {...navInfo}
          />
        ) : (
          <NavItem
            key={navInfo.link}
            menuIsMinimal={menuIsMinimal}
            {...navInfo}
          />
        ),
      )
    )
  }, [menuIsMinimal])

  const handlers = {
    pinNavbar: e => {
      updateMenuState({ variables: { expanded: !menuPinned } })
      setMenuPinned(!menuPinned)
    },
    mouseoverNav: e => menuIsMinimal && setMenuIsMinimal(false),
    mouseleaveNav: () => !menuPinned && setMenuIsMinimal(true),
    expandCollapse: () => setMenuIsMinimal(!menuIsMinimal),
  }

  return (
    <Root
      aria-expanded={!menuIsMinimal}
      className={className}
      data-testid="menu-container"
      onMouseLeave={handlers.mouseleaveNav}
    >
      <SectionNavLayoutSettings
        $pinned={menuPinned}
        onClick={handlers.pinNavbar}
        onMouseEnter={handlers.mouseoverNav}
        title={
          menuPinned
            ? t('menuSettings.MinimalSidebar')
            : t('menuSettings.KeepMenuVisible')
        }
        type="submit"
      >
        <StyledPinButton $menuIsMinimal={menuIsMinimal} $pinned={menuPinned} />
      </SectionNavLayoutSettings>
      {/* TODO: Place this notice (used for offline notification) better */}
      {notice}
      <MainNavWrapper>
        <UserComponent
          expanded={!menuIsMinimal}
          loginLink={loginLink}
          profileLink={profileLink}
          t={t}
          user={user}
        />
        <ScrollWrapper>
          <NavLinks>{renderLinks}</NavLinks>
        </ScrollWrapper>
      </MainNavWrapper>
    </Root>
  )
}

Menu.propTypes = {
  className: PropTypes.string.isRequired,
  loginLink: PropTypes.string.isRequired,
  navLinkComponents: PropTypes.arrayOf(PropTypes.object).isRequired, // eslint-disable-line react/forbid-prop-types
  user: PropTypes.oneOfType([PropTypes.object]),
  notice: PropTypes.node.isRequired,
  profileLink: PropTypes.string.isRequired,
}

Menu.defaultProps = {
  user: undefined,
}

export default Menu
