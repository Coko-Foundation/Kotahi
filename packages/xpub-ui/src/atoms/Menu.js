import React from 'react'
import classnames from 'classnames'
import classes from './Menu.local.css'

// TODO: match the width of the container to the width of the widest option?

class Menu extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      open: false,
      selected: props.value
    }
  }

  toggleMenu = () => {
    this.setState({
      open: !this.state.open
    })
  }

  handleSelect = selected => {
    this.setState({
      selected,
      open: false
    })

    this.props.onChange(selected)
  }

  optionLabel = value => {
    const { options } = this.props

    return options.find(option => option.value === value).label
  }

  render () {
    const { options, placeholder = 'Choose in the list' } = this.props
    const { open, selected } = this.state

    return (
      <div className={classnames(classes.root, {
        [classes.open]: open
      })}>
        <div>
          <button
            type="button"
            className={classes.opener}
            onClick={this.toggleMenu}>
            {selected ? (
              <span>{this.optionLabel(selected)}</span>
            ) : (
              <span className={classes.placeholder}>{placeholder}</span>
            )}
            <span className={classes.arrow}>â–¼</span>
          </button>
        </div>

        <div className={classes.menuContainer}>
          {open && (
            <div className={classes.menu}>
              {options.map(option => (
                <div
                  key={option.value}
                  className={classnames(classes.option, {
                    [classes.active]: option.value === selected
                  })}
                  onClick={() => this.handleSelect(option.value)}
                >{option.label || option.value}</div>
              ))}
            </div>
          )}
        </div>
      </div>
    )
  }
}

export default Menu
