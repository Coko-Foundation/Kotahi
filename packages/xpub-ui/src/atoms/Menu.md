A menu for selecting one of a list of options.

```js
const options = [
  { value: 'foo', label: 'Foo' },
  { value: 'bar', label: 'Bar' },
  { value: 'baz', label: 'Baz' }
];

<Menu 
  options={options}
  onChange={value => console.log(value)}/>
```

When an option is selected, it replaces the placeholder.

```js
const options = [
  { value: 'foo', label: 'Foo' },
  { value: 'bar', label: 'Bar' },
  { value: 'baz', label: 'Baz' }
];

<Menu 
  options={options}
  value="foo"
  onChange={value => console.log(value)}/>
```
