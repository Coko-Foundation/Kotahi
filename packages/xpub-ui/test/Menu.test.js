import React from 'react'
import { shallow } from 'enzyme'
// import renderer from 'react-test-renderer'

import Menu from '../src/atoms/Menu'

const props = {
  options: [
    { value: 'foo', label: 'Foo' },
    { value: 'bar', label: 'Bar' },
  ],
  value: 'foo'
}

const wrapper = shallow(<Menu {...props} />)

describe('Menu', () => {
  /*test('Snapshot', () => {
    const tree = renderer.create(
      <Menu {...props} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })*/

  test('Renders a Menu', () => {
    expect(wrapper.is('div')).toBeTruthy()
    expect(wrapper).toHaveLength(1)
  })
})
