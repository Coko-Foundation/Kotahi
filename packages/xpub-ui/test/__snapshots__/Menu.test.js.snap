// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Snapshot 1`] = `
<div
  className="root"
>
  <div
    className="main"
  >
    <div
      className="openerContainer"
    >
      <button
        className="opener"
        onClick={[Function]}
        type="button"
      >
        <span>
          Foo
        </span>
        <span
          className="arrow"
        >
          â–¼
        </span>
      </button>
    </div>
    <div
      className="optionsContainer"
    />
  </div>
</div>
`;
